hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/io/HoodieSparkMergeHandleV2.java

```diff
@@ -0,0 +1,413 @@
+      writeStatus.getStat().setFileId(fileId);
+      setWriteStatusPath();
+
+      // Create Marker file,
+      // uses name of `newFilePath` instead of `newFileName`
+      // in case the sub-class may roll over the file handle name.
+      createMarkerFile(partitionPath, newFilePath.getName());
+
+      // Create the writer for writing the new version file
+      fileWriter = HoodieFileWriterFactory.getFileWriter(instantTime, newFilePath, hoodieTable.getStorage(),
+          config, writeSchemaWithMetaFields, taskContextSupplier, HoodieRecord.HoodieRecordType.SPARK);
+    } catch (IOException io) {
+      LOG.error("Error in update task at commit " + instantTime, io);
+      writeStatus.setGlobalError(io);
+      throw new HoodieUpsertException("Failed to initialize HoodieUpdateHandle for FileId: " + fileId + " on commit "
+          + instantTime + " on path " + hoodieTable.getMetaClient().getBasePath(), io);
+    }
+  }
+
+  protected void setWriteStatusPath() {
+    writeStatus.getStat().setPath(new StoragePath(config.getBasePath()), newFilePath);
+  }
+
+  protected void makeOldAndNewFilePaths(String partitionPath, String oldFileName, String newFileName) {
+    oldFilePath = makeNewFilePath(partitionPath, oldFileName);
+    newFilePath = makeNewFilePath(partitionPath, newFileName);
+  }
+
+  // TODO(yihua): is this still needed
+  /*
+  if (needsUpdateLocation()) {
+        record.unseal();
+        record.setNewLocation(new HoodieRecordLocation(instantTime, fileId));
+        record.seal();
+      }
+   */
+
+  public void write() {
+    boolean usePosition = config.getBooleanOrDefault(MERGE_USE_RECORD_POSITIONS);
+    Schema readerSchema;
+    Option<InternalSchema> internalSchemaOption = Option.empty();
+    if (!StringUtils.isNullOrEmpty(config.getInternalSchema())) {
+      readerSchema = HoodieAvroUtils.addMetadataFields(
+          new Schema.Parser().parse(config.getSchema()), config.allowOperationMetadataField());
+      internalSchemaOption = SerDeHelper.fromJson(config.getInternalSchema());
+    } else {
+      readerSchema = HoodieAvroUtils.addMetadataFields(
+          new Schema.Parser().parse(config.getSchema()), config.allowOperationMetadataField());
+    }
+    // TODO(yihua): reader schema is good enough for writer?
+    HoodieFileGroupReader<T> fileGroupReader = new HoodieFileGroupReader<>(
+        readerContext,
+        storage.newInstance(hoodieTable.getMetaClient().getBasePath(), new HadoopStorageConfiguration(conf)),
+        hoodieTable.getMetaClient().getBasePath().toString(),
+        instantTime,
+        fileSlice,
+        readerSchema,
+        readerSchema,
+        internalSchemaOption,
+        hoodieTable.getMetaClient(),
+        hoodieTable.getMetaClient().getTableConfig().getProps(),
+        0,
+        Long.MAX_VALUE,
+        usePosition);
+    try {
+      fileGroupReader.initRecordIterators();
+      HoodieFileGroupReaderIterator<InternalRow> recordIterator
```

Comment:
does this also handle inserts sent to log files?... sth to handle

Reply:
Yes, inserts in log files are also handled as the basic feature in the file group reader (e.g., it supports compacting log file-only file slice).